### PR TITLE
Order yul subobjects by order of reference.

### LINKED ISSUE #15134: Different bytecode produced via IR for some of OpenZeppelin's contracts when extra contracts are included in the input
## Description

This looks like yet another bug where different AST IDs lead to differences in the generated bytecode. I found it when experimenting with parallel compilation of OpenZeppelin.

When each source file is compiled in isolation by providing only that single file as input (rather than providing all of them and only selecting one via `outputSelection`), three contracts produce different bytecode than when all contracts are compiled together:
- `test/metatx/ERC2771Forwarder.t.sol:ERC2771ForwarderTest`
- `test/token/ERC20/extensions/ERC4626.t.sol:ERC4626StdTest`
- `test/utils/structs/Checkpoints.t.sol:CheckpointsTrace224Test`

## Environment
- Compiler version: 0.8.26
- Target EVM version (as per compiler settings): cancun

## Steps to Reproduce
```bash
git clone --depth=1 https://github.com/OpenZeppelin/openzeppelin-contracts --branch v5.0.2
cd openzeppelin-contracts/
forge install

cat <<EOF > input.json
{
    "language": "Solidity",
    "sources": {
        "contracts/mocks/token/ERC4626Mock.sol":     {"urls": ["contracts/mocks/token/ERC4626Mock.sol"]},
        "test/token/ERC20/extensions/ERC4626.t.sol": {"urls": ["test/token/ERC20/extensions/ERC4626.t.sol"]}
    },
    "settings": {
        "remappings": [
            "@openzeppelin/contracts/=contracts/",
            "ds-test/=lib/forge-std/lib/ds-test/src/",
            "erc4626-tests/=lib/erc4626-tests/",
            "forge-std/=lib/forge-std/src/"
        ],
        "metadata": {"appendCBOR": false},
        "outputSelection": {"*": {"*": ["evm.bytecode"]}},
        "viaIR": true
    }
}
EOF

cat input.json \
    | solc --standard-json > output-select2.json
cat input.json \
    | jq 'del(.sources."contracts/mocks/token/ERC4626Mock.sol")' \
    | solc --standard-json > output-select1.json

diff --color --unified \
    <(cat output-select2.json \
        | jq '.contracts."test/token/ERC20/extensions/ERC4626.t.sol"."ERC4626StdTest".evm.bytecode.object' \
        | fold --width 160 \
    ) \
    <(cat output-select1.json \
        | jq '.contracts."test/token/ERC20/extensions/ERC4626.t.sol"."ERC4626StdTest".evm.bytecode.object' \
        | fold --width 160 \
    )
```

**Note**: This is a minimized example that's enough to reproduce one of the bytecode differences. With the full input there are more of them and they may or may not have the same cause. To make sure, when the bug is fixed, verify with the full original input: [`openzeppelin-5.0.2-full-input.json`](https://github.com/ethereum/solidity/files/15416204/openzeppelin-5.0.2-full-input.json). It should produce the same bytecode for the three contracts when compiled as is and when you include only the source files containing those contracts.

### PR DESCRIPTION
**Pull request by ekpyron**
Fixes https://github.com/ethereum/solidity/issues/15134

If someone is up for trying to craft a minimal test case for this, feel free to go ahead!

### CODE REVIEW

File: libsolidity/codegen/ir/IRGenerationContext.h [hunk 1]
```diff
@@ -143,7 +143,12 @@ class IRGenerationContext
 
 	RevertStrings revertStrings() const { return m_revertStrings; }
 
-	std::set<ContractDefinition const*, ASTNode::CompareByID>& subObjectsCreated() { return m_subObjects; }
+	std::vector<ContractDefinition const*> const& subObjectsCreated() { return m_subObjects; }
```
Comment (clonker) [inline libsolidity/codegen/ir/IRGenerationContext.h]: This could now be a const function I think

File: libsolidity/codegen/ir/IRGenerator.cpp [hunk 2]
```diff
@@ -101,7 +101,7 @@ std::string IRGenerator::generate(
 	std::map<ContractDefinition const*, std::string_view const> const& _otherYulSources
 )
 {
-	auto subObjectSources = [&_otherYulSources](std::set<ContractDefinition const*, ASTNode::CompareByID> const& subObjects) -> std::string
+	auto subObjectSources = [&_otherYulSources](util::UniqueVector<ContractDefinition const*> const& subObjects) -> std::string
```
Comment (cameel) [inline libsolidity/codegen/ir/IRGenerator.cpp]:
```suggestion
	auto subObjectSources = [&_otherYulSources](UniqueVector<ContractDefinition const*> const& _subObjects) -> std::string
```

File: libsolidity/codegen/ir/IRGenerationContext.h
Comment (cameel) [inline libsolidity/codegen/ir/IRGenerationContext.h:1]: We should have at least one test that's sensitive to the object order. Apparently none of the current tests is.

Something as simple as this seems to be enough to trigger this bug:
- `A.sol`:
    ```solidity
    contract A {}
    ```
- `B.sol`:
    ```solidity
    contract B {}
    ```
- `C.sol`:
    ```solidity
    import "A.sol";
    import "B.sol";

    contract C {
        A a = new A();
        B b = new B();
    }
    ```
```bash
solc B.sol C.sol --ir
```

On 0.8.26 it produces a different subobject order than `solc C.sol --ir` while on your branch it does not.

**Review by cameel — COMMENTED** [review]
The fix looks fine but I think we should cover it with at least a simple regression test before we merge.

File: libsolidity/codegen/ir/IRGenerationContext.h
Comment (ekpyron) [reply libsolidity/codegen/ir/IRGenerationContext.h:1]: One file is enough to test this even, I added a small case that shows the order change.

**Review by cameel — APPROVED** [review]

**Pull request merged**